### PR TITLE
Feat/overscanning

### PR DESCRIPTION
**Pull request by satyamgaur**
I want to discuss one thing, that I've used an additional param in `getVisibleVertices` and `getVisibleEdges`, `overscan`, with the help of which we can toggle the overscan feature if we want. And to implement overscanning for us I've declared another variable on _line 378_ inside the `renderChildren` function.. But I guess it would be much better if we would declare that variable in the constructor itself. But wanted to discuss it first before implementing.

### CODE REVIEW

File: .prettierrc [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "arrowParens": "avoid",
-  "printWidth": 120,
-  "singleQuote": true,
-  "trailingComma": "es5"
-}
+  "printWidth": 80,
```
Comment (smitpatel1010) [inline .prettierrc]: revert changes from this file

File: src/Diagram.js [hunk 2]
```diff
@@ -368,8 +375,9 @@ class Diagram extends React.PureComponent {
   }
 
   renderChildren(extremeX, extremeY, zoom) {
-    const verticesMap = this.getVisibleVertices(zoom);
-    const edges = this.getVisibleEdges(zoom);
+    const overscan = 1;
```
Comment (smitpatel1010) [inline src/Diagram.js]: lets take a prop `overscanPixels` as a number. and fallback of that should be width and height of the container

File: .prettierrc (hunk 1 above)
Comment (satyamgaur) [reply .prettierrc]: done

File: src/Diagram.js (hunk 2 above)
Comment (satyamgaur) [reply src/Diagram.js]: done

File: src/Diagram.js [hunk 3]
```diff
@@ -280,9 +252,16 @@ class Diagram extends React.PureComponent {
     const { width, height } = this.containerRef.current
       ? this.containerRef.current.getBoundingClientRect()
       : DEFAULT_CONTAINER_RECT;
+    const viewport = getViewport(scroll.left, scroll.top, width, height, zoom);
+    const viewportWithOverscanning = {
+      xMin: viewport.xMin - (this.props.overscanPixels ? this.props.overscanPixels.width : width),
```
Comment (milap1296) [inline src/Diagram.js]: this.props.overscanPixels this can be stored in a variable instead of doing it multiple times

Comment (satyamgaur) [reply src/Diagram.js]: will take care of it in the next commit :)

File: src/Diagram.js [hunk 4]
```diff
@@ -280,9 +252,16 @@ class Diagram extends React.PureComponent {
...
+      xMax: viewport.xMax + (this.props.overscanPixels ? this.props.overscanPixels.width : width),
+      yMin: viewport.yMin - (this.props.overscanPixels ? this.props.overscanPixels.height : height),
+      yMax: viewport.yMax + (this.props.overscanPixels ? this.props.overscanPixels.height : height),
```
Comment (smitpatel1010) [inline src/Diagram.js]: add `@babel/plugin-proposal-nullish-coalescing-operator` in `.babelrc` `plugins`
and then use `??`

Comment (satyamgaur) [reply src/Diagram.js]: done

Comment (smitpatel1010) [reply src/Diagram.js]: @satyamgaursprinklr need to add this in package.json as a devDependency too

File: package.json [hunk 5]
```diff
@@ -21,6 +21,7 @@
     "@babel/plugin-proposal-class-properties": "7.7.4",
     "@babel/plugin-proposal-optional-chaining": "^7.16.7",
     "@babel/plugin-transform-react-jsx-source": "7.7.4",
+    "@babel/plugin-proposal-nullish-coalescing-operator": "7.17.12",
```
Comment (smitpatel1010) [inline package.json:24]: Have you ran `yarn install` after adding in `package.json`?

Comment (satyamgaur) [reply package.json:24]: Oh, I wasn't aware of that, are there any more changes, I'll push all of them together in a single commit.

Comment (satyamgaur) [reply package.json:24]: done

**Pull request merged**